### DISCUSSION TITLE
[node] @types/node DOM Events is breaking Vue

### DISCUSSION THREAD

**Discussion by Jack-Works** [DefinitelyTyped/DefinitelyTyped#62759] [Issues with a @types package]
See the explanation at https://github.com/vuejs/core/issues/6899.

The breaking change is introduced by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59905

[a comment by an automated account is not shown]

**Comment by thw0rted** [comment]
Does [this suggestion](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62629#discussion_r993121000) fix the problem?  @OpportunityLiu suggested replacing `T` with `T['prototype']` -- I think your Vue issue actually helped me understand why that works.  You said you are trying to "reflect out" the "type side of a value" -- isn't that what `['prototype']` is doing?

**Reply by Jack-Works** [reply]
Yes, for this problem, `T['prototype']` works. It actually happened to be the type space `interface Event`.

**Reply by thw0rted** [reply]
Did you not get an error about `Event` having a recursive/circular reference to itself?  In a lib-dom environment, `__Event` should pass through to `globalThis.Event`, so `Event` extends `__Event` which extends `Event`, etc etc.  I'm at the point where I'm not sure how this ever worked.

**Reply by Jack-Works** [reply]
oh, when you get the correct `Event` interface via `globalThis.Event.prototype`, this trick fails because now it circularly inherits itself `Event (libdom) => __Event (types/node) => Event (libdom)`. Then I think this problem is impossible to solve. We need some help from the TypeScript team. cc @rbuckton and @sheetalkamat who reviewed this PR

**Reply by thw0rted** [reply]
I spent some time staring at this yesterday, and I finally realized that maybe I was trying to be too clever.  I think we just need

```ts
type __EventTarget = typeof globalThis extends { onmessage: any, EventTarget: any }
? {}
: { ... actual interface }
```

The circular reference is because we're trying to "re-augment" Event(Target) with itself.  Why are we doing that?  The conditional type is trying to "detect" if there's an existing global.  If that's the case, we actually just need to make sure we don't change it at all.

I'm not sure this is the right way to do that, but it seems to work. The idea is that extending `{}` should be a no-op.  Hopefully that's the canonical way to make `extends` do nothing?

**Reply by Jack-Works** [reply]
oh, that's correct! I also didn't think of that!

**Reply by thw0rted** [reply]
Forgot to share here, I just posted https://github.com/microsoft/DefinitelyTyped-tools/issues/546 which links to https://github.com/thw0rted/DefinitelyTyped/tree/node-global-interfaces-again.  The latter might solve the problem, but I had hoped to have a better answer to the dts-lint "meta-problem" first...

**Reply by falsandtru** [reply]
@thw0rted The fix of the bug you shipped is #62729. Can you merge and ship it?

**Reply by berniegp** [reply]
I'm also affected by this change when I implement `EventTarget`. [Repro here](https://www.typescriptlang.org/play?jsx=1&ts=4.8.4#code/PTAEAEBcGcFoFMAeAHeBjSCBOWD2XQAbeSUZQgQwE8BzPAVwDsATUAS0ekgsMIkiqpowRrmbwAUCHYBbZPlIDUoAN4BfUADM8M0AHJR4vVLABVRmlwyZ8RqUId4oCgCNcANyeRcoSFjY0NPAEkAAWTsF4WBISSPJYpOKaFPSEpGiU0NCgALJUAKKedrLk8DZ22YW2kAAqFFhBpCoSoM7MzFV2ADJsXLbBABQAlKpqMa3MvcgUkGihnZDDAFygbrjEFIyqoFgk9FhbfvTwANygYy07ZR7wCz19jIMj6hJjQA)

This code worked fine before
```
export default class MyEvent implements EventTarget {
  addEventListener() {}
  dispatchEvent(): boolean { return true; }
  removeEventListener() {}
}
```

it now produces this:
```
Class 'MyEvent' incorrectly implements interface 'EventTarget'.
  Property 'prototype' is missing in type 'MyEvent' but required in type 'EventTarget'.
```

I don`t know if this is the same root cause?

**Reply by thw0rted** [reply]
In case you missed it, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62782 should fix all related issues.  Please comment here if anyone is still having problems using the latest shipped version (I believe 18.11.2?)
